Refactored FSA reading code

chenguoguo opened the pull request:
This is WIP because I still have to test some real FSTs, e.g., lexicon.fst from Kaldi.

- [x] Separated reading code for K2 and OpenFST
- [x] Removed sorting but does not require source states in non-descending order for OpenFST style FSTs
- [x] Support omitted weights in OpenFST (treat it as 0.0f)
- [x] Test this for a real lexicon fst

@csukuangfj I added one argument `acceptor` (default true) for the Python fsa initialization OpenFST style WFSTs can have 1, 2, 4, 5 fields, which makes it difficult to figure acceptor/transducer automatically. Also, asking users to explicitly set acceptor/transducer may lead to less coding errors. But let me know your thoughts.

csukuangfj commented on k2/csrc/fsa_utils.cu:
```diff
@@ -99,75 +99,237 @@ static void SplitStringToVector(const std::string &in, const char *delim,
-// Create an acceptor from a stream.
-static Fsa AcceptorFromStream(std::string first_line, std::istringstream &is,
-                              bool openfst) {
+/* Create an acceptor from a stream, assuming the acceptor is in the K2 format:
```
In the original design without taking openfst into account, Dan wanted
a version accepting a parameter `negate_scores`. I am not sure
whether the parameter is still needed here.

chenguoguo replied:
@danpovey do we want to have the option for negating score when openfst == false? Or do we only want that when openfst == true?

csukuangfj commented on k2/csrc/fsa_utils.cu:
```diff
@@ -99,75 +99,237 @@ static void SplitStringToVector(const std::string &in, const char *delim,
+    } else {
+      K2_LOG(FATAL) << "Invalid line: " << line
+                    << "\nK2 transducer expects a line with 1 (final_state) or "
+                    << "4 (src_state dest_state label aux_label score) fields";
```
The output info is incorrect. 5 vs 4.

chenguoguo replied:
Good catch, fixed, thanks.

csukuangfj commented on k2/csrc/fsa_utils.cu:
```diff
@@ -99,75 +99,237 @@ static void SplitStringToVector(const std::string &in, const char *delim,
+
+   If there are multiple final states, we will create a super final state, and
+   create arcs from the old final states to the new super final state, with the
+   (negated) old final state cost/score as its cost/score, and -1 as its label.
```
0 as its aux_label.

chenguoguo replied:
We don't have aux_label for acceptors. I do have the comment for transducer, see here: https://github.com/k2-fsa/k2/blob/04872431289b5b3e9c154ef389d094d08a1a6e9c/k2/csrc/fsa_utils.cu#L356

csukuangfj commented on k2/csrc/fsa_utils.cu:
```diff
@@ -99,75 +99,237 @@ static void SplitStringToVector(const std::string &in, const char *delim,
-                        << "\nFinal state with weight detected in K2 format";
+      float score = 0.0f;
+      if (num_fields == 2u)
+        score = StringToFloat(splits[1]);
```
the scale -1 is missing.

chenguoguo replied:
The variable is called `original_final_weights` so I didn't negate it here. I did it later: https://github.com/k2-fsa/k2/blob/04872431289b5b3e9c154ef389d094d08a1a6e9c/k2/csrc/fsa_utils.cu#L316

But I can move it up if that's easier to read.

csukuangfj replied:
Please add a comment here or move it up. That's easier to read.

csukuangfj commented on k2/csrc/fsa_utils.cu:
```diff
@@ -99,75 +99,237 @@ static void SplitStringToVector(const std::string &in, const char *delim,
   std::vector<int32_t> original_final_states;
   std::vector<float> original_final_weights;
-  do {
+  while (std::getline(is, line)) {
```
we should either add a `bool finish` flag or check outside the loop
that the stream `is` is still in a good state.

chenguoguo replied:
For the openfst input, we will read till the end of the file since the final state may appear in the middle of the file. I added check for has_final_state as well as is.eof()

csukuangfj commented on k2/csrc/fsa_utils.cu:
```diff
@@ -99,75 +99,237 @@ static void SplitStringToVector(const std::string &in, const char *delim,
-    std::sort(arcs.begin(), arcs.end());
+
+  // Move arcs from "state_to_arcs" to "arcs".
+  std::size_t arc_index = 0;
```
`int32_t arc_index;` since `num_arcs` is of type `int32_t`

csukuangfj commented on k2/csrc/fsa_utils.cu:
```diff
@@ -99,75 +99,237 @@ static void SplitStringToVector(const std::string &in, const char *delim,
+    }
   }
+  K2_CHECK_EQ(num_arcs, static_cast<int32_t>(arc_index));
+  state_to_arcs.resize(0);
```
`resize(0)` will not free the underlying memory.
This statement can be deleted.

csukuangfj commented:
> I added one argument acceptor (default true) for the Python fsa initialization OpenFST style WFSTs can have 1, 2, 4, 5 fields, which makes it difficult to figure acceptor/transducer automatically

Can we define another python interface for reading OpenFST format? It will make the interface and user's life simpler.

chenguoguo commented:
Are you suggesting adding a new Python class, e.g., FsaOpenFst?

If yes, I'm inclined to not adding such a new class. The difference is really the reading, and once they are loaded, they will be equivalent to K2 Fsas. Having two classes Fsa and FsaOpenFst may make people feel they are different type of finite state acceptors.

But if you guys think it's a good idea to have a separate Python class for OpenFST, I can add that. Should be fairly straightforward.

csukuangfj commented:
```python
def read_openfst(...) -> Fsa:
   ...
```

chenguoguo commented:
I guess it makes sense to add that function. Will do that.

csukuangfj commented:
I would propose to leave the current interface as it is. To add support for a new format, we just add another interface.
A single interface to handle multiple formats makes things complicated.

chenguoguo commented:
So you'd like to keep the code to automatically figure out K2 FSAs/Transducers right?

csukuangfj commented:
> So you'd like to keep the code to automatically figure out K2 FSAs/Transducers right?

For the K2 part, I would say yes. Every field in the K2 format has to be present, so it is possible to figure out
automatically whether it is an acceptor or a transducer. I think the k2 format will be more popular than
OpenFST in K2, so a simpler interface for the k2 format makes life easier.

Not sure if @danpovey and @qindazhu have something to say about this.

csukuangfj commented:
I think you can leave your current C++ interface unchanged. The auto detection for the k2 format can be
supported in Python.

qindazhu commented:
> ```python
> def read_openfst(...) -> Fsa:
> ```

+1 ,
For k2 format, I prefer the approach as well which figures out FSA/FST automatically. For openFST format, I also would like to just provide a function to users and let them call the function to convert openFST format to k2 format.

csukuangfj commented:
If we are going to add I/O support for `FsaVec`, the extra support for OpenFst complicates things further.

chenguoguo commented:
OK sounds good. I'll make some changes later today when I get a chance.

danpovey commented:
Agreed: separate function for OpenFST I/O makes sense.

On Tue, Oct 6, 2020 at 7:40 AM Guoguo Chen <notifications@github.com> wrote:

> OK sounds good. I'll make some changes later today when I get a chance.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/k2-fsa/k2/pull/224#issuecomment-703946031>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZFLO5NE4VJPJ2W4QWQG5LSJJKQLANCNFSM4SEIDHJQ>
> .
>

danpovey commented:
... Incidentally, I did not have in mind reading FsaVec from text.  The
plan for reading Fsa and FsaVec efficiently is to use, say, PyTorch I/O to
read/write tensors of dimension N by 4 or N by 5 (for transducers) and to
use those to construct the FsaVec using functions I've already written (not
tested perhaps)... for writing you'd just store the arcs as a PyTorch
tensor of int32.  So when possible we'd use this fast binary I/O.

On Tue, Oct 6, 2020 at 12:08 PM Daniel Povey <dpovey@gmail.com> wrote:

> Agreed: separate function for OpenFST I/O makes sense.
>
> On Tue, Oct 6, 2020 at 7:40 AM Guoguo Chen <notifications@github.com>
> wrote:
>
>> OK sounds good. I'll make some changes later today when I get a chance.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/k2-fsa/k2/pull/224#issuecomment-703946031>, or
>> unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAZFLO5NE4VJPJ2W4QWQG5LSJJKQLANCNFSM4SEIDHJQ>
>> .
>>
>

chenguoguo commented:
Pushed more changes. Tested on Lexicon FST generated by Kaldi and worked. Removing "WIP"

danpovey commented on k2/csrc/fsa_utils.cu:
```diff
@@ -99,75 +99,241 @@ static void SplitStringToVector(const std::string &in, const char *delim,
+
+   @return It returns an Fsa on CPU.
+*/
+static Fsa OpenFSTAcceptorFromStream(std::istringstream &is) {
```
Google style guide says this should be OpenFst not OpenFST

danpovey commented on k2/python/k2/fsa.py:
```diff
@@ -18,7 +19,7 @@
 class Fsa(object):
 
-    def __init__(self, s: str, openfst: bool = False):
+    def __init__(self, s: str):
```
This is a preexisting issue, but I want to remind you guys that string input won't be the normal method; more commonly I anticpate reading a tensor in from disk, e.g. N by 4 or N by 5 of int32_t.

chenguoguo replied:
I was also wondering about that. What about we have
```
def from_string(...) -> 'Fsa':
```
for reading from strings, and let the default `__init__` function read from a tensor, something like
```
def __init__(self, tensor: torch.Tensor, aux_labels: Optional[torch.Tensor] = None) -> 'Fsa':
```

danpovey commented on k2/csrc/fsa_utils.cu:
```diff
@@ -99,75 +99,241 @@ static void SplitStringToVector(const std::string &in, const char *delim,
+      finished = true;               // set finish
+    } else {
+      K2_LOG(FATAL) << "Invalid line: " << line
+                    << "\nK2 acceptor expects a line with 1 (final_state) or "
```
BTW, we've been using lowercase k2 in text... this is open to discussion but I felt it would get complicated with directory names and variable names and the like if we consistently use uppercase.

danpovey commented on k2/csrc/fsa_utils.cu:
```diff
@@ -99,75 +99,241 @@ static void SplitStringToVector(const std::string &in, const char *delim,
+    }
+  }
+
+  K2_CHECK_EQ(finished, true) << "The last line should be the final state";
```
Actually it's not a requirement for OpenFST that the last line is a final state; also, in principle it's possible to have an FSA with no final states (although its not so useful).  So maybe remove this check?

danpovey commented on k2/csrc/fsa_utils.cu:
```diff
@@ -99,75 +99,241 @@ static void SplitStringToVector(const std::string &in, const char *delim,
     } else {
       K2_LOG(FATAL) << "Invalid line: " << line
-                    << "\nIt expects a line with 1, 2 or 4 fields";
+                    << "\nOpenFST acceptor expects a line with 1 (final_state),"
```
may not need the <<'s here..

chenguoguo commented:
I addressed most of the comments. For the `__init__` function of the `Fsa` class, I'd propose we make the default `__init__` function read from tensors, and have another function `from_string(...) -> Fsa` to read from strings. What do you guys think? If you think that would be a good change, I'll make some changes.

danpovey commented:
Thanks!
Yes, I agree init from tensors as default makes more sense.
If you want you can merge with my PR and test like that?  I haven't fully
tested that and today I'm not sure if I'll have time.



On Wed, Oct 7, 2020 at 2:39 AM Guoguo Chen <notifications@github.com> wrote:

> I addressed most of the comments. For the __init__ function of the Fsa
> class, I'd propose we make the default __init__ function read from
> tensors, and have another function from_string(...) -> Fsa to read from
> strings. What do you guys think? If you think that would be a good change,
> I'll make some changes.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/k2-fsa/k2/pull/224#issuecomment-704473795>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZFLO74VAAKNTJPJDJUMUDSJNP47ANCNFSM4SEIDHJQ>
> .
>

chenguoguo commented:
Sounds good, I'll find some time to make the change and test.

danpovey commented:
Thanks!!

On Wed, Oct 7, 2020 at 1:18 PM Guoguo Chen <notifications@github.com> wrote:

> Sounds good, I'll find some time to make the change and test.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/k2-fsa/k2/pull/224#issuecomment-704698125>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZFLO5E6Y6YE6XIDBWVA43SJP23XANCNFSM4SEIDHJQ>
> .
>

chenguoguo commented:
I was distracted by my other tasks so getting back to this late. But I have pushed changed to initialize Fsa() from a tensor, and also added some python test code for initializing from tensor.

danpovey commented on k2/csrc/fsa_utils.cu:
```diff
@@ -99,75 +99,241 @@ static void SplitStringToVector(const std::string &in, const char *delim,
+      int32_t src_state = StringToInt(splits[0]);
+      int32_t dest_state = StringToInt(splits[1]);
+      int32_t symbol = StringToInt(splits[2]);
+      float score = StringToFloat(splits[3]);
```
Not urgent, but we'll probably at some point get into  locale problems here, with comma vs. period for decimals.
 :-(  These things are hard to avoid.  I don't know whether the C code will behave the same w.r.t. locale as Python does.  In Kaldi we always recommended LC_ALL=C.  We should decide at some point what the best practice for this toolkit is.

chenguoguo replied:
I added a TODO to the StringToFloat function regarding the locale issue.

danpovey commented on k2/csrc/fsa_utils.cu:
```diff
@@ -177,101 +343,145 @@ static Fsa AcceptorFromStream(std::string first_line, std::istringstream &is,
+   We will negate the cost/score when we read them in. Also note, OpenFST may
+   omit the cost/score if it is 0.0.
+
+   If there are multiple final states, we will create a super final state, and
```
I don't believe it is actually conditional.
The code as you have it requires there to be at least one final state, no?
I think it would be better to always create a superfinal state, but not require there to be any final states.
(would just be disconnected in that case).

chenguoguo replied:
I thought about it for a second, but felt an FSA without a final state will be use less?

Let's do this then. We will always add a super final state. If there's at least one final state in the original FSA, we create an arc(s) from the original final state(s) to the super final state, with label -1; if there's no final state in the original FSA, we add the super final state (max_id + 1) but with no arcs connecting it with the original FSA. Or do you think we should add an arc from max_id to the super final state with label -1 and weight 0?

danpovey replied:
Always add super-final state but no special cases, we don't want to guess at what the user wants.  Better to have predictable behavior.  Non-coaccessible FSAs do sometimes arise naturally.

chenguoguo replied:
I removed the constrain of the presence of the final state in the OpenFST format. Super final state will always be added.

danpovey commented on k2/python/k2/fsa.py:
```diff
@@ -18,7 +19,48 @@
 class Fsa(object):
 
-    def __init__(self, s: str, openfst: bool = False):
+    def __init__(self, tensor: torch.Tensor,
```
OK... I notice we only have an Fsa object and not FsaVec.  FsaVec is actually going to be the more common case.  For now I think we can have both (not for this PR!) but we may eventually decide to unify them.

chenguoguo replied:
That's right, I planned to add FsaVec as a separate PR. But I'm not super clear yet whether or not I should add it as a separate class or just have one class. I'll create an issue for that and we can discuss from there. Once we have a conclusion, and can then implement it.

danpovey replied:
Yeah it's not clear to me either.  I kind of assumed Fangjun would be doing this part (FsaVec).  I will follow up with some other comments on the design.  In fact I'll create an issue.

csukuangfj replied:
I did not add `FsaVec` in Python because the construction of `FsaVec` had bugs which were fixed by #226 .

danpovey commented on k2/python/k2/fsa.py:
```diff
@@ -50,60 +92,58 @@ def __init__(self, s: str, openfst: bool = False):
-        '''
         ans = cls.__new__(cls)
         super(Fsa, ans).__init__()
+        fsa, aux_labels = _fsa_from_str(s, acceptor, False)
```
Instead of having aux_labels as a member directly, we may have it as one of the `per_arc` pseudo-members.
Basically, if we have an Fsa or FsaVec called f, we will have a bunch of fields like f.aux_labels, f.scores, f.foo, f.bar and so on.  They are not actually members but will correspond to keys in some kind of dict.  We'll have fairly general purpose code for propagating them; the main constraint is that the 1st index into them will correspond to the arc-index. 
In the specific case of aux_labels, it may end up getting treated specially for certain operations, like Invert() which would swap those labels with the `symbol` in the arc.

chenguoguo replied:
I'm not sure if I understand the pseudo-member part. Is this on the Python side, or on the C++ side accessed by Python through pybind? Do you have an example on this?

csukuangfj replied:
I think `pseudo-member` is via `__getattr__` and `__setattr__`, which is mentioned in one of the issues Dan just created.

danpovey commented:
Yes, pseudo-member would not be supported at the C++ level for now, only in
python.


On Fri, Oct 9, 2020 at 3:13 PM Fangjun Kuang <notifications@github.com>
wrote:

> *@csukuangfj* commented on this pull request.
> ------------------------------
>
> In k2/python/k2/fsa.py
> <https://github.com/k2-fsa/k2/pull/224#discussion_r502231364>:
>
> >          ans = cls.__new__(cls)
>          super(Fsa, ans).__init__()
> +        fsa, aux_labels = _fsa_from_str(s, acceptor, False)
>
> I think pseudo-member is via __getattr__ and __setattr__, which is
> mentioned in one of the issues Dan just created.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/k2-fsa/k2/pull/224#discussion_r502231364>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZFLOZVPJUYINHFVBU5BKDSJ2ZYFANCNFSM4SEIDHJQ>
> .
>

chenguoguo commented:
Oh I see issue #231 now. Looks like we will have a re-design on the k2.Fsa class. Shall we keep the aux_labels as is for now, and fix the pseudo-member issues when we re-do the k2.Fsa class in a separate PR?

danpovey commented:
Keeping them as-is is fine with me.

On Fri, Oct 9, 2020 at 3:37 PM Fangjun Kuang <notifications@github.com>
wrote:

> *@csukuangfj* commented on this pull request.
> ------------------------------
>
> In k2/python/k2/fsa.py
> <https://github.com/k2-fsa/k2/pull/224#discussion_r502242820>:
>
> > @@ -18,7 +19,48 @@
>
>  class Fsa(object):
>
> -    def __init__(self, s: str, openfst: bool = False):
> +    def __init__(self, tensor: torch.Tensor,
>
> I did not add FsaVec in Python because the construction of FsaVec had
> bugs which were fixed by #226 <https://github.com/k2-fsa/k2/pull/226> .
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/k2-fsa/k2/pull/224#discussion_r502242820>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZFLO4W5BKTQ34YRACKI73SJ24TVANCNFSM4SEIDHJQ>
> .
>
